From a22586305ddd191343cb27ad9c9061c93418ba07 Mon Sep 17 00:00:00 2001
From: Steve Dickson <SteveD@redhat.com>
Date: Thu, 18 Sep 2014 14:20:03 -0400
Subject: [fs] xprtrdma: Back off rkey when FAST_REG_MR fails

Message-id: <1411050057-12417-41-git-send-email-steved@redhat.com>
Patchwork-id: 96252
O-Subject: [RHEL7.1 PATCH 40/94] xprtrdma: Back off rkey when FAST_REG_MR fails
Bugzilla: 1111708
RH-Acked-by: J. Bruce Fields <bfields@redhat.com>
RH-Acked-by: Benjamin Coddington <bcodding@redhat.com>

From: Chuck Lever <chuck.lever@oracle.com>

If posting a FAST_REG_MR Work Reqeust fails, revert the rkey update
to avoid subsequent IB_WC_MW_BIND_ERR completions.

Suggested-by: Steve Wise <swise@opengridcomputing.com>
Signed-off-by: Chuck Lever <chuck.lever@oracle.com>
Signed-off-by: Anna Schumaker <Anna.Schumaker@Netapp.com>
Upstream Id: c93e986a295d537589efd0504f36ca952bd1a5be
Bugzilla: 1111708
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/sunrpc/xprtrdma/verbs.c b/net/sunrpc/xprtrdma/verbs.c
index dd1dabc..b670f4d 100644
--- a/net/sunrpc/xprtrdma/verbs.c
+++ b/net/sunrpc/xprtrdma/verbs.c
@@ -1577,6 +1577,7 @@ rpcrdma_register_frmr_external(struct rpcrdma_mr_seg *seg,
  if (rc) {
   dprintk("RPC:       %s: failed ib_post_send for register,"
    " status %i\n", __func__, rc);
+  ib_update_fast_reg_key(mr, --key);
   goto out_err;
  } else {
   seg1->mr_rkey = mr->rkey;
-- 
1.7.1